### PR TITLE
add repository resource, with tests

### PR DESCRIPTION
**Pull request by davidbloss**
## Issues

https://github.com/OpsLevel/team-platform/issues/308

## Changelog

[repository public docs](https://registry.terraform.io/providers/OpsLevel/opslevel/latest/docs/resources/repository)

Changes:
- Add `repository` resource, with tests

- [X] List your changes here
- [ ] Make a `changie` entry

### Using this file:
```tf
# main.tf
resource "opslevel_repository" "repo_with_alias" {
  identifier = "github.com:suraj-testing2/Shoes_Soda"
  owner      = "Z2lkOi8vb3BzbGV2ZWwvVGVhbS8yMjU"
}

resource "opslevel_repository" "repo_with_id" {
  identifier = "Z2lkOi8vb3BzbGV2ZWwvUmVwb3NpdG9yaWVzOjpHaXRodWIvNDYzMw"
  owner      = "Z2lkOi8vb3BzbGV2ZWwvVGVhbS8yMjU"
}
```

### Create Repository, `terraform apply`
```tf
Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # opslevel_repository.repo_with_alias will be created
  + resource "opslevel_repository" "repo_with_alias" {
      + id           = (known after apply)
      + identifier   = "github.com:suraj-testing2/Shoes_Soda"
      + last_updated = (known after apply)
      + owner        = "Z2lkOi8vb3BzbGV2ZWwvVGVhbS8yMjU"
    }

  # opslevel_repository.repo_with_id will be created
  + resource "opslevel_repository" "repo_with_id" {
      + id           = (known after apply)
      + identifier   = "Z2lkOi8vb3BzbGV2ZWwvUmVwb3NpdG9yaWVzOjpHaXRodWIvNDYzMw"
      + last_updated = (known after apply)
      + owner        = "Z2lkOi8vb3BzbGV2ZWwvVGVhbS8yMjU"
    }

Plan: 2 to add, 0 to change, 0 to destroy.
opslevel_repository.repo_with_id: Creating...
opslevel_repository.repo_with_alias: Creating...
opslevel_repository.repo_with_alias: Creation complete after 1s [id=Z2lkOi8vb3BzbGV2ZWwvUmVwb3NpdG9yaWVzOjpHaXRodWIvNDYzNA]
opslevel_repository.repo_with_id: Creation complete after 1s [id=Z2lkOi8vb3BzbGV2ZWwvUmVwb3NpdG9yaWVzOjpHaXRodWIvNDYzMw]

Apply complete! Resources: 2 added, 0 changed, 0 destroyed.
```

### Update this file:
```tf
# main.tf
resource "opslevel_repository" "repo_with_alias" {
  identifier = "github.com:suraj-testing2/Shoes_Soda"
  # owner      = "Z2lkOi8vb3BzbGV2ZWwvVGVhbS8yMjU"
}

resource "opslevel_repository" "repo_with_id" {
  identifier = "Z2lkOi8vb3BzbGV2ZWwvUmVwb3NpdG9yaWVzOjpHaXRodWIvNDYzMw"
  # owner      = "Z2lkOi8vb3BzbGV2ZWwvVGVhbS8yMjU"
}
```

### Update Repository, `terraform apply`
```tf
opslevel_repository.repo_with_id: Refreshing state... [id=Z2lkOi8vb3BzbGV2ZWwvUmVwb3NpdG9yaWVzOjpHaXRodWIvNDYzMw]
opslevel_repository.repo_with_alias: Refreshing state... [id=Z2lkOi8vb3BzbGV2ZWwvUmVwb3NpdG9yaWVzOjpHaXRodWIvNDYzNA]

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # opslevel_repository.repo_with_alias will be updated in-place
  ~ resource "opslevel_repository" "repo_with_alias" {
        id           = "Z2lkOi8vb3BzbGV2ZWwvUmVwb3NpdG9yaWVzOjpHaXRodWIvNDYzNA"
      + last_updated = (known after apply)
      - owner        = "Z2lkOi8vb3BzbGV2ZWwvVGVhbS8yMjU" -> null
        # (1 unchanged attribute hidden)
    }

  # opslevel_repository.repo_with_id will be updated in-place
  ~ resource "opslevel_repository" "repo_with_id" {
        id           = "Z2lkOi8vb3BzbGV2ZWwvUmVwb3NpdG9yaWVzOjpHaXRodWIvNDYzMw"
      + last_updated = (known after apply)
      - owner        = "Z2lkOi8vb3BzbGV2ZWwvVGVhbS8yMjU" -> null
        # (1 unchanged attribute hidden)
    }

Plan: 0 to add, 2 to change, 0 to destroy.
opslevel_repository.repo_with_id: Modifying... [id=Z2lkOi8vb3BzbGV2ZWwvUmVwb3NpdG9yaWVzOjpHaXRodWIvNDYzMw]
opslevel_repository.repo_with_alias: Modifying... [id=Z2lkOi8vb3BzbGV2ZWwvUmVwb3NpdG9yaWVzOjpHaXRodWIvNDYzNA]
opslevel_repository.repo_with_alias: Modifications complete after 1s [id=Z2lkOi8vb3BzbGV2ZWwvUmVwb3NpdG9yaWVzOjpHaXRodWIvNDYzNA]
opslevel_repository.repo_with_id: Modifications complete after 1s [id=Z2lkOi8vb3BzbGV2ZWwvUmVwb3NpdG9yaWVzOjpHaXRodWIvNDYzMw]

Apply complete! Resources: 0 added, 2 changed, 0 destroyed.
```

### Destroy Repository, `terraform destroy`
```tf
opslevel_repository.repo_with_alias: Refreshing state... [id=Z2lkOi8vb3BzbGV2ZWwvUmVwb3NpdG9yaWVzOjpHaXRodWIvNDYzNA]
opslevel_repository.repo_with_id: Refreshing state... [id=Z2lkOi8vb3BzbGV2ZWwvUmVwb3NpdG9yaWVzOjpHaXRodWIvNDYzMw]

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  - destroy

Terraform will perform the following actions:

  # opslevel_repository.repo_with_alias will be destroyed
  - resource "opslevel_repository" "repo_with_alias" {
      - id         = "Z2lkOi8vb3BzbGV2ZWwvUmVwb3NpdG9yaWVzOjpHaXRodWIvNDYzNA" -> null
      - identifier = "github.com:suraj-testing2/Shoes_Soda" -> null
    }

  # opslevel_repository.repo_with_id will be destroyed
  - resource "opslevel_repository" "repo_with_id" {
      - id         = "Z2lkOi8vb3BzbGV2ZWwvUmVwb3NpdG9yaWVzOjpHaXRodWIvNDYzMw" -> null
      - identifier = "Z2lkOi8vb3BzbGV2ZWwvUmVwb3NpdG9yaWVzOjpHaXRodWIvNDYzMw" -> null
    }

Plan: 0 to add, 0 to change, 2 to destroy.
opslevel_repository.repo_with_alias: Destroying... [id=Z2lkOi8vb3BzbGV2ZWwvUmVwb3NpdG9yaWVzOjpHaXRodWIvNDYzNA]
opslevel_repository.repo_with_id: Destroying... [id=Z2lkOi8vb3BzbGV2ZWwvUmVwb3NpdG9yaWVzOjpHaXRodWIvNDYzMw]
opslevel_repository.repo_with_id: Destruction complete after 0s
opslevel_repository.repo_with_alias: Destruction complete after 0s

Destroy complete! Resources: 2 destroyed.
```

### CODE REVIEW

File: opslevel/resource_opslevel_repository.go [hunk 1]
```diff
@@ -1,99 +1,264 @@
...
+		Attributes: map[string]schema.Attribute{
+			"id": schema.StringAttribute{
+				Description: "The ID of the repository.",
+				Computed:    true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.UseStateForUnknown(),
+				},
+			},
+			"identifier": schema.StringAttribute{
+				Description: "The id or human-friendly, unique identifier for the repository.",
```
Comment (taimoorgit) [inline opslevel/resource_opslevel_repository.go:60]: Overall I find this field (and the old version too) pretty confusing - shouldn't this be called `alias`? I don't think that an ID can be passed here. 🤔

File: tests/resources.tf [hunk 2]
```diff
@@ -63,6 +63,13 @@ resource "opslevel_infrastructure" "big_infra" {
   schema = "Big Database"
 }
 
+# Repository resources
+
+resource "opslevel_repository" "mock_repo" {
```
Comment (taimoorgit) [inline tests/resources.tf:68]: Recommend adding a test where owner is not set

**Review by taimoorgit — APPROVED** [review]
This is a really confusing resource, LGTM so long as tophatting works.

File: opslevel/resource_opslevel_repository.go (hunk 1 above)
Comment (davidbloss) [reply opslevel/resource_opslevel_repository.go:64]: ID is used in one spot down below to get the repo in the create operation.
	`repository, err = r.client.GetRepository(*opslevel.NewID(identifier))`

**Pull request merged**